Common/include/matrix_structure.inl

```diff
@@ -160,25 +175,364 @@ template<class ScalarType>
       
       break;
     }
   }
   
 }
 
+template<class ScalarType>
+inline ScalarType *CSysMatrix<ScalarType>::GetBlock_ILUMatrix(unsigned long block_i, unsigned long block_j) {
+
+  for (unsigned long index = row_ptr_ilu[block_i]; index < row_ptr_ilu[block_i+1]; index++)
+    if (col_ind_ilu[index] == block_j)
+      return &(ILU_matrix[index*nVar*nEqn]);
+
+  return NULL;
+}
+
+template<class ScalarType>
+inline void CSysMatrix<ScalarType>::SetBlock_ILUMatrix(unsigned long block_i, unsigned long block_j, ScalarType *val_block) {
+  
+  unsigned long iVar, index;
+  
+  for (index = row_ptr_ilu[block_i]; index < row_ptr_ilu[block_i+1]; index++) {
+    if (col_ind_ilu[index] == block_j) {
+      for (iVar = 0; iVar < nVar*nEqn; iVar++)
+        ILU_matrix[index*nVar*nEqn+iVar] = val_block[iVar];
+      break;
+    }
+  }
+  
+}
+
+template<class ScalarType>
+inline void CSysMatrix<ScalarType>::SetBlockTransposed_ILUMatrix(unsigned long block_i, unsigned long block_j, ScalarType *val_block) {
+
+  unsigned long iVar, jVar, index;
+
+  for (index = row_ptr_ilu[block_i]; index < row_ptr_ilu[block_i+1]; index++) {
+    if (col_ind_ilu[index] == block_j) {
+      for (iVar = 0; iVar < nVar; iVar++)
+        for (jVar = 0; jVar < nEqn; jVar++)
+          ILU_matrix[index*nVar*nEqn+iVar*nEqn+jVar] = val_block[jVar*nVar+iVar];
+      break;
+    }
+  }
+
+}
+
+template<class ScalarType>
+inline void CSysMatrix<ScalarType>::SubtractBlock_ILUMatrix(unsigned long block_i, unsigned long block_j, ScalarType *val_block) {
+  
+  for (unsigned long index = row_ptr_ilu[block_i]; index < row_ptr_ilu[block_i+1]; index++) {
+    if (col_ind_ilu[index] == block_j) {
+      MatrixSubtraction(&ILU_matrix[index*nVar*nEqn], val_block, &ILU_matrix[index*nVar*nEqn]);
+      break;
+    }
+  }
+  
+}
+
+template<class T, bool alpha, bool beta, bool transp>
+inline void gemv_impl(const unsigned long n, const T *a, const T *b, T *c) {
+  /*---
+   This is a templated version of GEMV with the constants as boolean
+   template parameters so that they can be optimized away at compilation.
+   This is still the traditional "row dot vector" method.
+  ---*/
+  unsigned long i, j;
+  for (i = 0; i < n; i++) {
+    if (!beta) c[i] = 0.0;
+    for (j = 0; j < n; j++) {
+      if (alpha) c[transp? j:i] += a[i*n+j] * b[transp? i:j];
+      else       c[transp? j:i] -= a[i*n+j] * b[transp? i:j];
+    }
+  }
+}
+
+template<class T>
+inline void gemm_impl(const unsigned long n, const T *a, const T *b, T *c) {
+  /*--- Same deal as for GEMV but here only the type is templated. ---*/
+  unsigned long i, j, k;
+  for (i = 0; i < n; i++) {
+    for (j = 0; j < n; j++) {
+      c[i*n+j] = 0.0;
+      for (k = 0; k < n; k++)
+        c[i*n+j] += a[i*n+k] * b[k*n+j];
+    }
+  }
+}
+
+#define __MATVECPROD_SIGNATURE__(TYPE,NAME) \
+inline void CSysMatrix<TYPE>::NAME(const TYPE *matrix, const TYPE *vector, TYPE *product)
```

Comment:
I hope this does not look too bad, but this MKL stuff does work very well for the discrete adjoint (maybe 10% faster than native code).